### PR TITLE
IMPRO-1838 manipulate reliability table CLI

### PR DESCRIPTION
**Pull request by fionaRust**
* Added a new test to apply_reliability_calibration CLI. This test takes a cubelist as an input, which requires new clize functionality. It takes a cubelist as an input as the reliability table may have different lengths for the probability bin coordinate for different thresholds after the ManipulateReliabilityTable plugin. This means the reliability tables for the different thresholds can't be merged into a single cube and the ManipulateReliabilityTable plugin outputs a cubelist rather than  a cube. The ApplyReliabilityCalibration CLI needs to be able to read in the cubelist output from ManipulateReliabilityTable plugin.
* Added a new CLI for the ManipulateReliabilityTable plugin and associated tests.

Testing:
 - [ ] Ran tests and they passed OK
 - [ ] Added new tests for the new feature(s)

CLA
 - [ ] If a new developer, signed up to CLA

### CODE REVIEW

File: improver_tests/cli/test_init.py [hunk 1]
```diff
@@ -135,6 +135,17 @@ def test_basic(self, m):
         self.assertEqual(result, "return")
 
 
+class Test_inputcubelist(unittest.TestCase):
+    """Tests the input cubelist function"""
+
+    @patch("improver.cli.maybe_coerce_with", return_value="return")
+    def test_basic(self, m):
+        """Tests that input cubelist calls load_cubelist with the string"""
+        result = inputcubelist("foo")
```
Comment (benfitzpatrick) [inline improver_tests/cli/test_init.py:144]: Codacy found an issue: [Undefined variable 'inputcubelist'](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6105320)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6105320&bid=19984885)

**Review by neilCrosswaite — APPROVED** [review]
An `inputcubelist` is probably to right way to do this.

Looks good to me.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/manipulate_reliability_table.py  1
- improver_tests/acceptance/test_manipulate_reliability_table.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6105320&bid=19984885)

File: improver_tests/acceptance/test_manipulate_reliability_table.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import pytest
+
+from . import acceptance as acc
+
+pytestmark = [pytest.mark.acc, acc.skip_if_kgo_missing]
+CLI = acc.cli_name_with_dashes(__file__)
+run_cli = acc.run_cli(CLI)
+
+
+def test_mainpulate(tmp_path):
```
Comment (gavinevans) [inline improver_tests/acceptance/test_manipulate_reliability_table.py:42]:
```suggestion
def test_manipulate(tmp_path):
```

File: improver_tests/acceptance/test_manipulate_reliability_table.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    """
+    Test mainpulation of a reliablity table
```
Comment (gavinevans) [inline improver_tests/acceptance/test_manipulate_reliability_table.py:44]:
```suggestion
    Test manipulation of a reliability table
```

File: improver_tests/acceptance/test_manipulate_reliability_table.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    table_path = kgo_dir / "reliability_table_precip.nc"
+    output_path = tmp_path / "output.nc"
+    args = [table_path, "--output", output_path]
+    run_cli(args)
+    acc.compare(output_path, kgo_path)
+
+
+def test_mainpulate_minimum_forecast_count(tmp_path):
+    """
+    Test mainpulation of a reliablity table with a reduced minimum forecast count
```
Comment (gavinevans) [inline improver_tests/acceptance/test_manipulate_reliability_table.py:57]:
```suggestion
def test_manipulate_minimum_forecast_count(tmp_path):
    """
    Test manipulation of a reliability table with a reduced minimum forecast count
```

File: improver/cli/manipulate_reliability_table.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""CLI to mainpulate a relibility table cube."""
```
Comment (gavinevans) [inline improver/cli/manipulate_reliability_table.py:32]:
```suggestion
"""CLI to manipulate a reliability table cube."""
```

File: improver/cli/manipulate_reliability_table.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+
+@cli.clizefy
+@cli.with_output
+def process(
+    reliability_table: cli.inputcube, *, minimum_forecast_count: int = 200,
+):
+    """
+    Manipulate a reliability table to ensure sufficient sample counts in
+    as many bins as possible by combining bins low sample counts.
```
Comment (gavinevans) [inline improver/cli/manipulate_reliability_table.py:44]:
```suggestion
    as many bins as possible by combining bins with low sample counts.
```

File: improver/cli/manipulate_reliability_table.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            spatial dimensions have been aggregated.
+        minimum_forecast_count (int):
+            The minimum number of forecast counts in a forecast probability
+            bin for it to be used in calibration.
+            The default value of 200 is that used in Flowerdew 2014.
+
+    Returns:
+        iris.cube.CubeList:
+            The reliability table that has been manipulated to ensure
+            sufficient sample counts and monotonic observation frequency.
```
Comment (gavinevans) [inline improver/cli/manipulate_reliability_table.py:59]:
```suggestion
            sufficient sample counts in each probability bin and a monotonic observation frequency.
```

File: improver_tests/acceptance/test_manipulate_reliability_table.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    kgo_dir = acc.kgo_root() / "manipulate-reliability-table/basic"
+    kgo_path = kgo_dir / "kgo_precip.nc"
+    table_path = kgo_dir / "reliability_table_precip.nc"
+    output_path = tmp_path / "output.nc"
+    args = [table_path, "--output", output_path]
+    run_cli(args)
+    acc.compare(output_path, kgo_path)
+
+
+def test_mainpulate_minimum_forecast_count(tmp_path):
```
Comment (gavinevans) [inline improver_tests/acceptance/test_manipulate_reliability_table.py:55]:
```suggestion
def test_manipulate_minimum_forecast_count(tmp_path):
```

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks for adding this CLI, @fionaRust 👍 

This looks fine to me. I've added a few typo and docstring corrections.

File: improver/cli/apply_reliability_calibration.py [hunk 9]
```diff
@@ -37,7 +37,7 @@
 @cli.clizefy
 @cli.with_output
 def process(
-    forecast: cli.inputcube, reliability_table: cli.inputcube = None,
+    forecast: cli.inputcube, reliability_table: cli.inputcubelist = None,
```
Comment (bayliffe) [inline improver/cli/apply_reliability_calibration.py:40]: The doc-string below should be modified to indicate that reliability_table can be a cube or cubelist.

File: improver_tests/acceptance/test_manipulate_reliability_table.py (hunk 2 above)
Comment (bayliffe) [reply improver_tests/acceptance/test_manipulate_reliability_table.py:42]: @gavinevans you're really against mainpulating, even though it's what all the cool kids are doing.

File: improver_tests/acceptance/test_manipulate_reliability_table.py (hunk 4 above)
Comment (bayliffe) [reply improver_tests/acceptance/test_manipulate_reliability_table.py:57]: Isn't this an increased minimum forecast count, rather than decreased?

**Review by bayliffe — CHANGES_REQUESTED** [review]
A couple of tiny things.

**Review by bayliffe — APPROVED** [review]

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/acceptance/test_manipulate_reliability_table.py  1
- improver/cli/manipulate_reliability_table.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6105320&bid=19984885)

**Review by gavinevans — APPROVED** [review]

**Pull request merged**